source/client/HandAndFootButtons.py

```diff
@@ -0,0 +1,165 @@
+
+It is Hand and Foot specific because different games may need additional buttons, and may need them arranged
+ differently, but most of this code should be useful for other games.
+"""
+
+
+def CreateButtons(self):
+    """This creates the buttons used for HandAndFoot. """
+    self.ready_yes_btn = Btn.Button(UIC.White, (UIC.Disp_Width-150), (UIC.Disp_Height-70), 125, 25, text='Ready:YES')
+    self.ready_color_idx = 2  # color of outline will be: UIC.outline_colors(ready_color_idx)
+    self.ready_no_btn = Btn.Button(UIC.White, (UIC.Disp_Width-150), (UIC.Disp_Height-30), 125, 25, text='Ready:NO')
+    self.not_ready_color_idx = 6  # color of outline will be: UIC.outline_colors(ready_color_idx)
+    self.sort_status_btn = Btn.Button(UIC.White, 900, 25, 225, 25, text='sort by status')
+    self.sort_btn = Btn.Button(UIC.White, 900, 75, 225, 25, text='sort by number')
+    self.prepare_card_btn = Btn.Button(UIC.White, 400, 25, 345, 25, text='Selected cards -> prepared cards')
+    self.clear_prepared_cards_btn = Btn.Button(UIC.White, 320, 75, 225, 25, text='Clear prepared cards')
+    self.play_prepared_cards_btn = Btn.Button(UIC.White, 600, 75, 225, 25, text='Play prepared cards')
+    self.discard_action_btn = Btn.Button(UIC.Bright_Red, 190, 25, 100, 25, text='discard')
+    return
+
+def ButtonDisplay(self):
+    # display draw pile and various action buttons
+    loc_xy = (self.draw_pile.x, self.draw_pile.y)
+    self.draw_pile.draw(self.display, loc_xy, self.draw_pile.outline_color)
+    # update discard info and redraw
+    discard_info = self.controller.getDiscardInfo()
+    self.top_discard = discard_info[0]
+    self.pickup_pile_sz = discard_info[1]
+    if self.pickup_pile_sz > 0:
+        self.top_discard_wrapped = UICardWrapper(self.top_discard, (100, 25), UIC.scale)
+        self.pickup_pile = self.top_discard_wrapped.img_clickable
+        loc_xy = (self.pickup_pile.x, self.pickup_pile.y)
+        self.pickup_pile.draw(self.display, loc_xy, self.pickup_pile.outline_color)
+        self.labelMedium(str(self.pickup_pile_sz), 150, 35)
+    if self.controller._state.round == -1:
+        self.ready_yes_btn.draw(self.display, self.ready_yes_btn.outline_color)
+        self.ready_no_btn.draw(self.display, self.ready_no_btn.outline_color)
+    self.sort_status_btn.draw(self.display, self.sort_status_btn.outline_color)
+    self.sort_btn.draw(self.display, self.sort_btn.outline_color)
+    self.prepare_card_btn.draw(self.display, self.prepare_card_btn.outline_color)
+    self.clear_prepared_cards_btn.draw(self.display, self.clear_prepared_cards_btn.outline_color)
+    self.play_prepared_cards_btn.draw(self.display, self.play_prepared_cards_btn.outline_color)
+    self.discard_action_btn.draw(self.display, self.discard_action_btn.outline_color)
+    return
+
+
+def ClickedButton(self, pos):
+    # carry out action after mouse clicked on button.
+    if self.pickup_pile_sz > 0:
```

Comment:
I'm not sure what pickup_pile_sz is. If its the size of the discard pile then there is a check in the controller, if its a boolean for 'does the discard pile exist at all' then theoretically an empty discard pile of size 0 should be handled by the controller properly and we probably don't need to track that here?

Reply:
self.pickup_pile_sz = discard_info[1] 
 where discard_info = self.controller.getDiscardInfo().
If pickup_pile_sz is 0 and I don't have this check, then program crashes when it tries to create pick-up pile button (there's no wrapped card to draw as the top card on the pile). It's also used to label the size of the pile (the small blue number that appears on top of the pile).